crmd/notify.c

```diff
@@ -20,120 +20,501 @@
+                nano_digits = (nano_digits > 6)?6:nano_digits;
+                nano_digits = (nano_digits < 0)?0:nano_digits;
+                sprintf(&nanofmt_s[1], ".%ds", nano_digits);               
+            } else {
+                if (format[fmt_pos+fmt_len] != 0) {
+                    continue;
+                }
+                fmt_pos = scanned_pos; /* print till end */
+            }
+        } else {
+            scanned_pos = strlen(format);
+            fmt_pos = scanned_pos; /* print till end */
+        }
+        tmp_fmt_s = strndup(&format[printed_pos], fmt_pos - printed_pos);
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wformat-nonliteral"
+        date_len += strftime(&date_s[date_len], max-date_len, tmp_fmt_s, &tm);
+#pragma GCC diagnostic pop
+        printed_pos = scanned_pos;
+        free(tmp_fmt_s);
+        if (nano_digits) {
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wformat-nonliteral"
+            date_len += snprintf(&date_s[date_len], max-date_len, 
+                                 nanofmt_s, nano_s);
+#pragma GCC diagnostic pop 
+            nano_digits = 0;
+        }
+    }
+
+    return (date_len == 0)?NULL:strdup(date_s);
+}
+
+/*
+ * end of possibly generic time-handling stuff
+ */
+
+
+/*
+ * syncronize local data with cib
+ */
+
+static void free_notify_list_entry(notify_entry_t *entry)
+{
+    free(entry->id);
+    free(entry->path);
+    free(entry->tstamp_format);
+    free(entry->info);
+    free(entry->recipient);
+    free(entry);
+}
+
+static void free_notify_list()
+{
+    if (notify_list) {
+        g_list_free_full(notify_list, (GDestroyNotify) free_notify_list_entry);
+        notify_list = NULL;
+    }
+}
+
+static void add_dup_notify_list_entry(notify_entry_t *entry)
+{
+    notify_entry_t *new_entry = 
+        (notify_entry_t *) calloc(1, sizeof(notify_entry_t));
+
+    CRM_ASSERT(new_entry);
+    *new_entry = (notify_entry_t) {
+       .id = strdup(entry->id),
+       .path = strdup(entry->path),
+       .timeout = entry->timeout,
+       .tstamp_format = entry->tstamp_format?strdup(entry->tstamp_format):NULL,
+       .info = entry->info?strdup(entry->info):NULL,
+       .recipient = entry->recipient?strdup(entry->recipient):NULL
+    };
+    notify_list = g_list_append(notify_list, new_entry);
+}
+
+void
+notifications_query_callback(xmlNode * msg, int call_id, int rc, 
+                             xmlNode * output, void *user_data)
+{
+    xmlNode *notify = output;
+    notify_entry_t entry;
+
+    free_notify_list();
+
+    if (notify_script) {
+        entry = (notify_entry_t) {
+            .id = notify_script,
+            .path = notify_script,
+            .timeout = CRMD_NOTIFY_DEFAULT_TIMEOUT_MS,
+            .recipient = notify_target
+        };
+        add_dup_notify_list_entry(&entry);
+        crm_notice("Legacy Notifications enabled");
+    } else {
+        crm_notice("Legacy Notifications disabled");
+    }
```

Comment:
I'm thinking the configuration should be either-or: if any new-style alerts are configured, ignore the legacy options. The legacy compatibility is only to preserve the operation of configurations that existed before the new syntax. Once someone switches to the new syntax, any leftover old syntax should be ignored (we could log a message if that's the case).

XML IDs can't contain slashes, so it's probably best to use a static string for id in the legacy case, since there can only be one script configured.

Don't bother logging a message if legacy options aren't in use; the proportion of configurations using them should be tiny, and there's no need to increase the logs of everyone else.